microceph/ceph/osd.go

```diff
@@ -438,28 +439,413 @@ func AddOSD(s *state.State, path string, wipe bool, encrypt bool) error {
+	logger.Debugf("Reweighting osd.%d to %f", osd, weight)
+	_, err := processExec.RunCommand(
+		"ceph", "osd", "crush", "reweight",
+		fmt.Sprintf("osd.%d", osd),
+		fmt.Sprintf("%f", weight),
+	)
+	if err != nil {
+		// only log a warn, don't treat fail to reweight as a fatal error
+		logger.Warnf("Failed to reweight osd.%d: %v", osd, err)
+	}
+}
+
+func doPurge(osd int64) error {
+	// run ceph osd purge command
+	_, err := processExec.RunCommand(
+		"ceph", "osd", "purge", fmt.Sprintf("osd.%d", osd),
+		"--yes-i-really-mean-it",
+	)
+	return err
+}
+
+func purgeOSD(osd int64) error {
+	var err error
+	retries := 10
+	var backoff time.Duration
+
+	for i := 0; i < retries; i++ {
+		err = doPurge(osd)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// we're getting a RunError from processExec.RunCommand, and it
+		// wraps the original exit error if there's one
+		exitError, ok := err.(shared.RunError).Unwrap().(*exec.ExitError)
+		if !ok {
+			// not an exit error, abort and bubble up the error
+			logger.Warnf("Purge failed with non-exit error: %v", err)
+			break
+		}
+		if syscall.Errno(exitError.ExitCode()) != syscall.EBUSY {
+			// not a busy error, abort and bubble up the error
+			logger.Warnf("Purge failed with unexpected exit error: %v", exitError)
+			break
+		}
+		// purge failed with EBUSY - retry after a delay, and make delay exponential
+		logger.Infof("Purge failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
+		time.Sleep(backoff)
+	}
+
+	if err != nil {
+		logger.Errorf("Failed to purge osd.%d: %v", osd, err)
+		return fmt.Errorf("Failed to purge osd.%d: %w", osd, err)
+	}
+	logger.Infof("osd.%d purged", osd)
+	return nil
+}
+
+func wipeDevice(s common.StateInterface, osd int64) error {
+	// get the device path
+	path, err := database.OSDQuery.Path(s.ClusterState(), osd)
+	if err != nil {
+		return err
+	}
+	// wipe the device, retry with exponential backoff
+	retries := 8
+	var backoff time.Duration
+	for i := 0; i < retries; i++ {
+		err = timeoutWipe(path)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// wipe failed - retry after a delay, and make delay exponential
+		logger.Infof("Wipe failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
+		time.Sleep(backoff)
+	}
+	if err != nil {
+		// log a warning, but don't treat wipe failure as a fatal error
+		// e.g. if the device is broken, we still want to remove it from the cluster
+		logger.Warnf("Fault during device wipe: %v", err)
+	}
+	return nil
+}
+
+// timeoutWipe wipes the given device with a timeout, in order not to hang on broken disks
+func timeoutWipe(path string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	_, err := processExec.RunCommandContext(
+		ctx,
+		"dd", "if=/dev/zero",
```

Comment:
I know this is unrelated to this change, but I'm not really a fan of this way of wiping a disk. I much prefer `wipefs -a -f ${path}` since it can catch some occasions where the disk may appear 'broken', but that's just a personal preference.

Reply:
I've thought about going with wipefs but the downside is you need to pull in a new package into the snap, whereas `dd` is already on board, that's why I opted to keep it